Builder/TreeBuilder.php

```diff
@@ -0,0 +1,246 @@
+
+
+/**
+ * Builds a tree from a list of routes
+ */
+class TreeBuilder
+{
+    const ROUTE_OPTIONS_KEY = "page_tree";
+
+    /**
+     * All configuration options.
+     *
+     * Mapping of routing configuration to the properties of the Node
+     * ["config-option" => "propertyPath"]
+     *
+     * @var string[]
+     */
+    private static $configurationOptions = [
+        "title" => "title",
+        "hidden" => "hidden",
+        "separator" => "separator",
+        "parameters" => "parameters",
+        "security" => "security",
+        "parent" => "parentRoute",
+    ];
+
+
+    /**
+     * @var PropertyAccessor
+     */
+    private $propertyAccessor;
+
+
+    /**
+     * @var ParametersGenerator
+     */
+    private $parametersGenerator;
+
+
+
+    /**
+     * @param ParametersGenerator $parametersGenerator
+     */
+    public function __construct (ParametersGenerator $parametersGenerator)
+    {
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+        $this->parametersGenerator = $parametersGenerator;
+    }
+
+
+
+    /**
+     * Builds the route tree
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     *
+     * @return Node[] the array is indexed by route name
+     */
+    public function buildTree ($routeCollection)
+    {
+        $relevantRoutes = $this->calculateRelevantRoutes($routeCollection);
+        $nodes = $this->generateNodesFromRoutes($routeCollection, $relevantRoutes);
+        $nodes = $this->linkNodeHierarchy($nodes);
+
+        // needs to be after the hierarchy has been set up
+        return $this->calculateAllParameters($nodes);
+    }
+
+
+
+    /**
+     * Calculates a list of which routes should be included in the tree
+     *
+     * @param Route[]|RouteCollection $routeCollection
+     *
+     * @return array of format ["route" => (bool) $includeInTree]
+     * @throws InvalidRouteTreeException
+     */
+    private function calculateRelevantRoutes ($routeCollection)
+    {
+        $routeIndex = [];
+
+        foreach ($routeCollection as $routeName => $route)
+        {
+            $routeIndex[$routeName] = false;
+        }
+
+        foreach ($routeCollection as $routeName => $route)
+        {
+            $routeData = $route->getOption(self::ROUTE_OPTIONS_KEY);
+
+            // no route data found -> skip
+            if (null === $routeData)
+            {
+                continue;
+            }
+
+            // route data found -> mark as relevant
+            $routeIndex[$routeName] = true;
+
+            // mark parent route as relevant
+            if (isset($routeData["parent"]) && !empty($routeData["parent"]))
+            {
+                $parentRoute = $routeData["parent"];
+
+                if (!isset($routeIndex[$parentRoute]))
+                {
+                    throw new InvalidRouteTreeException(sprintf(
+                        "Route '%s' references a parent '%s', but the parent route does not exist.",
```

Comment:
Any reason for not using string interpolation here?

Nitpick: I'd rather read something like `"..., but the parent route could not be found."`.